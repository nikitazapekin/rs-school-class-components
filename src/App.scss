.redirect__btn {
	position: fixed;
	right: 0;
	bottom: 0;
	padding: 10px;
	margin-bottom: 100px;
	font-size: 32px;
}
