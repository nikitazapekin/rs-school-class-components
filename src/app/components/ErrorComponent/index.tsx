import React, { useState } from 'react';
import styles from "./index.module.scss";

const ErrorComponent: React.FC = () => {
	const [counter, setCounter] = useState<number>(0);

	const handleClick = () => {
		setCounter(counter + 1);
	};

	if (counter === 1) {
		throw new Error('I crashed!!');
	}

	return (
		<div>
			<button className={styles.throw__error} onClick={handleClick} data-testid="throw__error">
				Throw Error
			</button>
		</div>
	);
};

export default ErrorComponent;
