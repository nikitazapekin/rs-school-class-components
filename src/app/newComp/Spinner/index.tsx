//import './index.scss';
import styles from "./index.module.scss"
const Spinner = () => {
	return (
		<>
			<span className={styles.loader} data-testid="loader"></span>
		</>
	);
};
export default Spinner;
