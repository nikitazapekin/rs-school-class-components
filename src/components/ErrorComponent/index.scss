.throw__error {
    padding: 20px 25px;
    cursor: pointer;
    font-size: 33px;
  position: fixed;
  right: 0;
  bottom: 0;
}