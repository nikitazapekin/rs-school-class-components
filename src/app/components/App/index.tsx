 "use client"
 import styles from "./index.module.scss"
 import Card from "../Card";
 import React from "react";
import { useAppDispatch } from "../../../hooks/redux";
import { useSelector } from "react-redux";
import { paramsSelector } from "../../../redux/selectors/getSearchParams";
import useURL from "../../../hooks/useURL";
import { setPrevPageActionCreator } from "../../../redux/action-creators/setSearchParamsActionCreator";
import { storedUsersSelector } from "../../../redux/selectors/getStoredElements";
import { setNextPageActionCreator } from "../../../redux/action-creators/setSearchParamsActionCreator";
import StoredUsersButton from "../StoredUsersFlyoutElement";
import ErrorComponent from "../ErrorComponent";
import { ErrorBoundary } from "../ErrorBoundary";
import Providers from "../../../redux/Provider";
type UserDataArray = Array<{
	login: string;
	id: number;
	node_id: string;
	avatar_url: string;
	gravatar_id: string;
	url: string;
	html_url: string;
	followers_url: string;
	following_url: string;
	gists_url: string;
	starred_url: string;
	subscriptions_url: string;
	organizations_url: string;
	repos_url: string;
	events_url: string;
	received_events_url: string;
	type: string;
	site_admin: boolean;
	score: number;
}>; 

interface AppProps {
	users: UserDataArray;
  }
function App({ users }: AppProps) {
 const dispatch = useAppDispatch()
const params = useSelector(paramsSelector)
const {setPage} =useURL()
const storedUsers = useSelector(storedUsersSelector);
	const handleNext = () => {
		window.scrollTo(0, 0);
		dispatch(setNextPageActionCreator());
		setPage(params.offset + 1, params.query);
	};

	const handlePrev = () => {
		window.scrollTo(0, 0);
		dispatch(setPrevPageActionCreator());
		setPage(params.offset - 1, params.query);
	};
  return (
  <>
<Providers>
<ErrorBoundary>
<section className={styles.list}>
			<div className={styles.list__container}>
				{users.map((item) => (
					<Card user={item} key={item.id} />
					))}
				{users.length === 0 && <p className={styles.list__nothing}>Nothing found</p>}
				<div className={styles.list__btns}>
					<button className={`${styles.list__prev} ${styles.list__btn}`} onClick={handlePrev}
					>
						Prev
					</button>
					<button className={`${styles.list__next} ${styles.list__btn}`} onClick={handleNext} 
					data-testid="next">
						Next
					</button>
				</div>
			</div>
		</section>
		{storedUsers.length > 0 && <StoredUsersButton />}
 <ErrorComponent />
 
	</ErrorBoundary>
						</Providers>
  </>
  );
  }
  export default  App 