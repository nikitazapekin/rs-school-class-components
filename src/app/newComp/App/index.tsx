 "use client"
 import styles from "./index.module.scss"
 import Card from "../Card";
 
import { useAppDispatch } from "../../../hooks/redux";
import { useSelector } from "react-redux";
import { paramsSelector } from "../../../redux/selectors/getSearchParams";
import useURL from "../../../hooks/useURL";
import { setPrevPageActionCreator } from "../../../redux/action-creators/setSearchParamsActionCreator";
import { storedUsersSelector } from "../../../redux/selectors/getStoredElements";
import { setNextPageActionCreator } from "../../../redux/action-creators/setSearchParamsActionCreator";
import StoredUsersButton from "../StoredUsersFlyoutElement";
import ErrorComponent from "../ErrorComponent";
import { ErrorBoundary } from "../ErrorBoundary";
import Providers from "../../../redux/Provider";
type UserDataArray = Array<{
	login: string;
	id: number;
	node_id: string;
	avatar_url: string;
	gravatar_id: string;
	url: string;
	html_url: string;
	followers_url: string;
	following_url: string;
	gists_url: string;
	starred_url: string;
	subscriptions_url: string;
	organizations_url: string;
	repos_url: string;
	events_url: string;
	received_events_url: string;
	type: string;
	site_admin: boolean;
	score: number;
}>; 

interface AppProps {
	users: UserDataArray;
  }
function App({ users }: AppProps) {
 const dispatch = useAppDispatch()
const params = useSelector(paramsSelector)
const {setPage} =useURL()
const storedUsers = useSelector(storedUsersSelector);
	const handleNext = () => {
		window.scrollTo(0, 0);
		dispatch(setNextPageActionCreator());
		setPage(params.offset + 1, params.query);
	};

	const handlePrev = () => {
		window.scrollTo(0, 0);
		dispatch(setPrevPageActionCreator());
		setPage(params.offset - 1, params.query);
	};
  return (
  <>
<Providers>

<ErrorBoundary>

<section className={styles.list}>
			<div className={styles.list__container}>
			{/*}	{isLoading && <Spinner />} */}
	 
				{users.map((item) => (
					<Card user={item} key={item.id} />
					))}
				{users.length === 0 && <p className={styles.list__nothing}>Nothing found</p>}
				<div className={styles.list__btns}>
					<button className={`${styles.list__prev} ${styles.list__btn}`} onClick={handlePrev}
					>
						Prev
					</button>
					<button className={`${styles.list__next} ${styles.list__btn}`} onClick={handleNext} 
					data-testid="next">
						Next
					</button>
				</div>
			</div>
		</section>
		{storedUsers.length > 0 && <StoredUsersButton />}
 <ErrorComponent />
 
	</ErrorBoundary>
						</Providers>
  </>
  );
  }
  export default  App
  /* 

  import Spinner from '../Spinner';
import Card from '../Card';
import styles from "./index.module.scss";
import { useAppDispatch } from '../../../hooks/redux';
import { useSelector } from 'react-redux';
import { setNextPageActionCreator, setPrevPageActionCreator } from '../../../redux/action-creators/setSearchParamsActionCreator';
import { paramsSelector } from '../../../redux/selectors/getSearchParams';
import { isLoadingSelector } from '../../../redux/selectors/isLoadingSelector';
import useURL from '../../../hooks/useURL';
import { UserItem } from './types';
import { getUsersSelector } from '../../../redux/selectors/getUsersSelector';
import { getUserData } from '../../../api';
import { useEffect } from 'react';
import { setUsersActionCreator } from '../../../redux/action-creators/setUsersActionCreator';
import { ThemeProvider } from '../ThemeContext';
const List = () => {
	const { setPage } = useURL();
	const dispatch = useAppDispatch();
	const users = useSelector(getUsersSelector);
	const params = useSelector(paramsSelector);
 
	const isLoading = useSelector(isLoadingSelector);

	const handleNext = () => {
		window.scrollTo(0, 0);
		dispatch(setNextPageActionCreator());
		setPage(params.offset + 1, params.query);
	};

	const handlePrev = () => {
		window.scrollTo(0, 0);
		dispatch(setPrevPageActionCreator());
		setPage(params.offset - 1, params.query);
	};
 
useEffect(() => {
	const fetchData = async () => {
		try {
			console.log("FETCHING", params.limit, params.offset, params.query )
			const data = await getUserData(params.limit, params.offset, params.query);
			dispatch(setUsersActionCreator(data))
	 
		} catch (err) {
	 
		}  
	};
	fetchData();
}, [params.limit, params.offset, params.query]); 
	return (
		<ThemeProvider>

		<section className={styles.list}>
			<div className={styles.list__container}>
				{isLoading && <Spinner />}
				{users.map((item: UserItem) => (
					<Card user={item} key={item.id} />
				))}
				{users.length === 0 && <p className={styles.list__nothing}>Nothing found</p>}
				<div className={styles.list__btns}>
					<button className={`${styles.list__prev} ${styles.list__btn}`} onClick={handlePrev}>
						Prev
					</button>
					<button className={`${styles.list__next} ${styles.list__btn}`} onClick={handleNext} data-testid="next">
						Next
					</button>
				</div>
			</div>
		</section>
		</ThemeProvider>
	);
};

export default List;


*/
/*

*/
/*
  	const { query, page , limit = '10' } = context.query;
	const offset = parseInt(page as string, 10);
	const perPage = parseInt(limit as string, 10);
	let url: string;
	if (!query) {
		url = `https://api.github.com/search/users?q=type:user&page=${offset}&per_page=${perPage}`;
	} else {
		url = `https://api.github.com/search/users?q=${query}&page=${offset}&per_page=${perPage}`;
	}
	try {
		let users: UserDataArray = [];

		if (!query) {
			const response: AxiosResponse<UserData> = await axios.get(url);
			users = response.data.items;
		} else {
			const response: AxiosResponse<UserData> = await axios.get(url);
			users = response.data.items;
		}
		return {
			props: {
				users,
				},
		};
	} catch (error) {
		if (axios.isAxiosError(error)) {
			const axiosError = error as AxiosError;
			console.error('Error fetching data:', axiosError.message);
		} else {
			console.error('Unknown error:', error);
		}
		return {
			props: {
				users: [],
			 
				},
		};
	}
    */





/*
import axios, { AxiosError, AxiosResponse } from 'axios';

type UserDataArray = Array<{
	login: string;
	id: number;
	node_id: string;
	avatar_url: string;
	gravatar_id: string;
	url: string;
	html_url: string;
	followers_url: string;
	following_url: string;
	gists_url: string;
	starred_url: string;
	subscriptions_url: string;
	organizations_url: string;
	repos_url: string;
	events_url: string;
	received_events_url: string;
	type: string;
	site_admin: boolean;
	score: number;
}>;
interface UserData {
	total_count: number;
	incomplete_results: boolean;
	items: UserDataArray;
}

async function fetchUserData(url: string): Promise<UserDataArray> {
	try {
		const response: AxiosResponse<UserDataArray> = await axios.get(url);
		return response.data;
	} catch (error) {
		if (axios.isAxiosError(error)) {
			const axiosError = error as AxiosError;
			console.error('Error fetching data:', axiosError.message);
		} else {
			console.error('Unknown error:', error);
		}
		throw new Error('Error fetching data');
	}
}
async function fetchUserDataAdvanced(url: string): Promise<UserData> {
	try {
		const response: AxiosResponse<UserData> = await axios.get(url);
		return response.data;
	} catch (error) {
		if (axios.isAxiosError(error)) {
			const axiosError = error as AxiosError;
			console.error('Error fetching data:', axiosError.message);
		} else {
			console.error('Unknown error:', error);
		}
		throw new Error('Error fetching data');
	}
}
  */








/*

import axios from 'axios';

export default async function App() {
  const response = await axios.get('https://jsonplaceholder.typicode.com/posts');
  const posts = response.data;

  return (
    <ul>
      {posts.map((post) => (
        <li key={post.id}>{post.title}</li>
      ))}
    </ul>
  );
}
  */