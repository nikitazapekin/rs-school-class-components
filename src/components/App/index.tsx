import Header from '../Header';
import List from '../List';
import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { storedUsersSelector } from '../../store/selectors/getStoredElements';
import StoredUsersButton from '../StoredUsersFlyoutElement';
import Background from '../Background';
import { useAppDispatch } from '../../hooks/redux';
import { setStoredInLocalStorageActionCreator } from '../../store/action-creators/setStoredInLocalStorageQuery';
import { isLoadingSelector } from '../../store/selectors/isLoadingSelector';
import Spinner from '../Spinner';

import { useNavigation } from '@remix-run/react';
 

const App = () => {
	const dispatch = useAppDispatch();
	const storedUsers = useSelector(storedUsersSelector);
	const isLoading = useSelector(isLoadingSelector);
	useEffect(() => {
		dispatch(
			setStoredInLocalStorageActionCreator(
				String(localStorage.getItem('searchParam') != null ? localStorage.getItem('searchParam') : ''),
			),
		);
	}, []);

	//const { state } = useNavigation();
	//console.log(state)
	return (
		<>
		
{/*
		</>
		)}
 {state === 'loading' ? (
    <div>
    loading
    </div>
    ) : (
      <>
    
			{state === 'loading' && (
    <div>
    loading
    </div>
			)
			}
		*/}
			<div className="container">
			<Header />
				<List />
				{storedUsers.length > 0 && <StoredUsersButton />}

			</div>
			<Background />

			
			{/*
			{isLoading && <Spinner />}
			*/}
		</>
	);
};

export default App;
 