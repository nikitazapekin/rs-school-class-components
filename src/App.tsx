
import Header from './components/Header';
import List from './components/List';
import { ErrorBoundary } from './components/ErrorBoundary';

import { useEffect} from 'react';
import { useSelector } from 'react-redux';
import { storedUsersSelector } from './store/selectors/getStoredElements';
import StoredUsersButton from './components/StoredUsersFlyoutElement';
import { ThemeProvider } from './components/ThemeContext';
import Background from './components/Background';
import ErrorComponent from './components/ErrorComponent';
 
 
import { useAppDispatch } from './hooks/redux';
import { setStoredInLocalStorageActionCreator } from './store/action-creators/setStoredInLocalStorageQuery';
import { isLoadingSelector } from './store/selectors/isLoadingSelector';
import Spinner from './components/Spinner';
 
const App = () => {
  const dispatch = useAppDispatch()
  const storedUsers = useSelector(storedUsersSelector);
  const isLoading = useSelector(isLoadingSelector)

 
  useEffect(() => {
    dispatch(setStoredInLocalStorageActionCreator(String(localStorage.getItem("searchParam") != null ? localStorage.getItem("searchParam") : "")))
  }, [])
  return (
    <>
      <ErrorBoundary>
        <ThemeProvider>
          <div className="container">
            <Header />
            <List
            />
            {storedUsers.length > 0 && (
              <StoredUsersButton />
            )}
            <ErrorBoundary>
              <ErrorComponent />
            </ErrorBoundary>
          </div>
          <Background />
           {isLoading && <Spinner />}
        </ThemeProvider>
      </ErrorBoundary>
    </>
  );
};

export default App;

 /*
 import { MemoryRouter } from "react-router-dom";
import App from "./App";
import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { rootReducer } from "./store/store";
import { RootState } from "./store/store";

// Define the initial state according to the expected structure
const initialState: Partial<RootState> = {
  appSlice: {
    isLoading: false,
    error: null,
    isLoadingUserData: false,
    users: [],
    params: {
      limit: 10,
      offset: 1,
      query: "",
      storedValue: ""
    }
  },
  selectedElementsSlice: {
    storedElements: [],
    selectedElement: null
  },
   
};

const store = configureStore({
  reducer: rootReducer,
  preloadedState: initialState
});

describe("TodoItem", () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

*/