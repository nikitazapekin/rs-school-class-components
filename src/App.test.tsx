import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { isLoadingSelector } from './redux/selectors/isLoadingSelector';
import { storedUsersSelector } from './redux/selectors/getStoredElements';
import App from './App';
import List from './components/List';
import ErrorComponent from './components/ErrorComponent';
import StoredUsersButton from './components/StoredUsersFlyoutElement';
import Spinner from './components/Spinner';
jest.mock('./components/List');
jest.mock('./components/ErrorComponent');
jest.mock('./components/StoredUsersFlyoutElement');
jest.mock('./components/Spinner');
jest.mock('./redux/selectors/getStoredElements', () => ({
  storedUsersSelector: jest.fn(),
}));
jest.mock('./redux/selectors/isLoadingSelector', () => ({
  isLoadingSelector: jest.fn(),
}));


const mockStore = () => {
  return configureStore({
    reducer: {
    },
    preloadedState: {
     
    },
  });
};

describe('App component', () => {
  beforeEach(() => {
    (List as jest.Mock).mockImplementation(() => <div>List Component</div>);
    (ErrorComponent as jest.Mock).mockImplementation(() => <div>Error Component</div>);
    (StoredUsersButton as jest.Mock).mockImplementation(() => <div>Stored Users Button</div>);
    (Spinner as jest.Mock).mockImplementation(() => <div>Loading...</div>);
  });

  it('renders the list and error components', () => {
    (storedUsersSelector as jest.Mock).mockReturnValue([]);
    (isLoadingSelector as jest.Mock).mockReturnValue(false);

    render(
      <Provider store={mockStore()}>
        <App />
      </Provider>
    );

    expect(screen.getByText('List Component')).toBeInTheDocument();
    expect(screen.getByText('Error Component')).toBeInTheDocument();
  });

  it('shows the stored users button when there are stored users', () => {
    (storedUsersSelector as jest.Mock).mockReturnValue([{}]);
    (isLoadingSelector as jest.Mock).mockReturnValue(false);

    render(
      <Provider store={mockStore()}>
        <App />
      </Provider>
    );

    expect(screen.getByText('Stored Users Button')).toBeInTheDocument();
  });

  it('shows the spinner when loading', () => {
    (storedUsersSelector as jest.Mock).mockReturnValue([]);
    (isLoadingSelector as jest.Mock).mockReturnValue(true);

    render(
      <Provider store={mockStore()}>
        <App />
      </Provider>
    );

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });
});
 