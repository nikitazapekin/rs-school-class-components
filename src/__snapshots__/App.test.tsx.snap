// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <div
      class="error__overlay"
      data-testid="error__overlay"
    />
    <div
      class="error__inner"
    >
      <p
        class="error__title"
      >
        Something went wrong
      </p>
      <button
        class="error__btn"
        data-testid="error__btn"
      >
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;
